### DISCUSSION TITLE
harbor.http.register.simple not working

### DISCUSSION THREAD

**Discussion by redimongo** [savonet/liquidsoap#3911] [User Support]
Ok I have just installed a new copy of liquidsoap for ubuntu which seems to be running Liquidsoap 2.0.2  note sure why when I did ```apt install liquidsoap```

I kinda have it working, however I am gettign a really strange issue:
At test1.liq, line 29, char 0-20:
harbor.http.register.simple(

Error 5: this value has no method simple
```
def redirect_icecast(request, response) =
  response.redirect("http://139.59.108.222:9008#{request.path}")
end

# Register this handler at port 8005 (provided harbor sources are also served
# from this port).
harbor.http.register.simple(
  port=8005,
  method="GET",
  handler=redirect_icecast
)

```
I also tried the code available at https://www.liquidsoap.info/blog/2023-03-10-how-to-run-a-http-server-with-liquidsoap/ with no success,

At server.liq, line 6, char 2-13:
  harbor.http.response(

Error 5: this value has no method response
```
def handler(request) =
  log("Got a request on path #{request.path}, protocol version: #{request.http_version}, \
       method: #{request.method}, headers: #{request.headers}, query: #{request.query}, \
       body: #{request.body()}")

  harbor.http.response(
    content_type="text/html",
    data="<p>ok, this works!</p>"
  )
end

harbor.http.register.simple(port=8080, method="GET", path, handler)
```

**Comment by vitoyucepi** [comment]
Hi @redimongo,
You are using liquidsoap from the ubuntu 22.04 repository. Since then, there have been two releases with major breaking changes: 2.1 and 2.2. You can either use the documentation for https://www.liquidsoap.info/doc-2.0.7/, which will work most of the time, or upgrade to the latest version of liquidsoap.
Here's the installation guide https://www.liquidsoap.info/doc-2.2.4/install.html.

To use the latest version of liquidsoap, or any other, you can
1. Use a container image if you're familiar with docker or podman.
   For example `savonet/liquidsoap:v2.2.5`.
1. Install from release artifacts on github.
   https://github.com/savonet/liquidsoap/releases/tag/v2.2.5.
1. Build from opam, but this requires some special knowledge.
1. Finally, upgrade to ubuntu 24.04. I can't recommend this at the moment, but I should mention it as a solution.

**Reply by redimongo** [reply]
I decided to upgrade the server to 24.04 - fingers crossed it's a fresh install so should be ok.

**Reply by vitoyucepi** [reply]
If I were you, I would prefer either 1 or 2, because liquidsoap *2.2.5*, which is currently the latest release, is only available in containers or for ubuntu *22.04*.
If you upgrade to ubuntu *24.04* your options will be limited to 1 or 3 if you want to use the most recent features.

**Reply by redimongo** [reply]
I agree docker is better, but I am just trying to get proof of concept. The proof of concept is instead of redirecting users like this
```
def redirect_icecast(request, response) =

 if request.path == "/stream" then
        response.redirect("http://139.59.108.222:8003#{request.path}")
    end

end
```

It should just be able to connect to the stream - I don't know if that is even possible.

Also update went well and was successful.

**Comment by redimongo** [comment]
issue has been fixed with upgrade.

**Comment by gAlleb** [comment]
Yeah! Cool!

@vitoyucepi, hi, I've thought I could catch you here. Probably my mind isn't in a good state cause I'm trying to resolve the issue but can't. Simple explanation could be useful. So:

```
services:
  liquidsoap:
    image: savonet/liquidsoap:v2.2.5
    command:
      - /home/radio/t.liq
    volumes:
      - ./:/home/radio


settings.log.level := 3
s = playlist("/home/radio/playlist.m3u", mode="normal", reload_mode="watch")
output.dummy(s, fallible=true)
```

Works in ubuntu 22.04.

But it refuses to work in `24.04` - `Docker version 26.1.1, build 4cf5afa`:

```
[+] Running 2/2
 ✔ Network radio_default         Created                                   0.3s 
 ✔ Container radio-liquidsoap-1  Created                                   0.1s 
Attaching to liquidsoap-1
liquidsoap-1  | At line 1, char 0:
liquidsoap-1  | 
liquidsoap-1  | Error 3: Syntax error!
liquidsoap-1 exited with code 1
```

What am I missing?

Docker has been installed via official apt repository as per `https://docs.docker.com/engine/install/ubuntu/`

**Reply by vitoyucepi** [reply]
You're trying to use Docker Compose volumes on a remote host, aren't you?
If you want to mount something to the remote Docker container from localhost, you could:
1. Build a new image with files.
1. Use environment variables.
1. Use Docker in swarm mode.
1. Use something like Kubernetes.

For single host installations, I'm not a big fan of either swarm or kubes. So I prefer to maintain my own container build infrastructure, define environment variables, and in-place builds using docker compose.

**Reply by gAlleb** [reply]
nonono. Ok

```
services:
  liquidsoap:
    image: savonet/liquidsoap:v2.2.5
    command:
      - /home/radio/t.liq
    volumes:
      - ./:/home/radio


s = sine() 
output(s)
```

Just doesn't work in ubuntu 24.04.

```[+] Running 2/2
 ✔ Network radio_default         Created                                   0.3s 
 ✔ Container radio-liquidsoap-1  Created                                   0.1s 
Attaching to liquidsoap-1
liquidsoap-1  | At line 1, char 0:
liquidsoap-1  | 
liquidsoap-1  | Error 3: Syntax error!
liquidsoap-1 exited with code 1
```

Fresh install, docker. Same works in 22.04

**Reply by gAlleb** [reply]
It blows my mind actually) Works everywhere except 24.04

Tried other dockerized containers in 24.04 (not LS) - they all work fine

**Reply by vitoyucepi** [reply]
I can't confirm this on my end.
Try to create an interactive container.
`docker run -it --rm -v "$(pwd):/home/radio" -w /home/radio --entrypoint bash savonet/liquidsoap:v2.2.5`.

Are you sure you are running on localhost? Try `docker context ls` to see if it is using a unix socket or local ip.
By the way, the latest version of docker is *26.1.2*.

**Reply by gAlleb** [reply]
```
NAME        DESCRIPTION                               DOCKER ENDPOINT               ERROR
default *   Current DOCKER_HOST based configuration   unix:///var/run/docker.sock   
```

Ok. It's some kind of permission error I have never encountered before.

If i `docker compose up` in `/tmp` directory with `t.liq` inside of it (owner: radio) - LS starts.
If i `docker compose up` in `/home/radio` directory with `t.liq` inside of it (owner: radio) - LS doesn't start.

```
:/home/radio$ docker run -it --rm -v "$(pwd):/home/radio" -w /home/radio --entrypoint bash savonet/liquidsoap:v2.2.5
liquidsoap@7979dc5baa53:/home/radio$ ls
ls: cannot open directory '.': Permission denied
```

```
:/tmp$ docker run -it --rm -v "$(pwd):/home/radio" -w /home/radio --entrypoint bash savonet/liquidsoap:v2.2.5
liquidsoap@33570b3fda95:/home/radio$ ls
docker-compose.yaml
liqaa01catmp
snap-private-tmp
systemd-private-c20e71d1536d47b5a89a622d74582d01-ModemManager.service-LGS2aL
systemd-private-c20e71d1536d47b5a89a622d74582d01-apache2.service-JcJsgZ
systemd-private-c20e71d1536d47b5a89a622d74582d01-polkit.service-QXtJMb
systemd-private-c20e71d1536d47b5a89a622d74582d01-systemd-logind.service-CpAnKg
systemd-private-c20e71d1536d47b5a89a622d74582d01-systemd-resolved.service-d09aD7
systemd-private-c20e71d1536d47b5a89a622d74582d01-systemd-timesyncd.service-RoCrmu
t.liq
```

I have never encountered this before.

Why it works in 22.04?
The setup is identical

UPDATE. Yeah - I'm starting to understand that shit)
```
services:
  liquidsoap:
    image: savonet/liquidsoap:v2.2.5
    command:
      - /tmp/t.liq
    volumes:
      - /tmp:/tmp
```
docker-compose.yaml is in /home/radio and LS starts OK.

100% my bad. 

2 questions. How to make it it work with `/home/radio/` directory? Why it's working in `22.04` already without any modifications?

**Reply by vitoyucepi** [reply]
It's very possible that by binding the directory into the Docker container, you are passing acls and permissions. So `ls -al /home/radio` or `stat /home/radio` inside and outside the container.

If you want to mount the directory in the container, you will need to configure permissions, like `0777` of `/tmp`, or fine-tune the filesystem acl.

**Reply by gAlleb** [reply]
```
radio@sxvx:~$ docker run -it --rm -v "$(pwd):/home/radio" -w /home/radio --entrypoint bash savonet/liquidsoap:v2.2.5
liquidsoap@9a6de8e87431:/home/radio$ ls -l
ls: cannot open directory '.': Permission denied
liquidsoap@9a6de8e87431:/home/radio$ exit
exit
radio@sxvx:~$ ls -l
total 36
-rwxrwxr-x 1 radio radio  135 May 10 16:15 docker-compose.yaml
drwxrwxr-x 8 radio radio 4096 May  7 21:07 liquidsoap
drwxr-xr-x 3 radio radio 4096 May  7 16:34 music
drwxrwxr-x 3 radio radio 4096 Apr 29 21:06 packages
drwxrwxr-x 2 radio radio 4096 May  7 22:54 playlist
drwxrwxr-x 4 radio radio 4096 Apr 12 17:07 scripts
drwxrwxr-x 3 radio radio 4096 May  7 21:55 sources
drwxrwxr-x 2 radio radio 4096 Feb  5 00:26 testaudio
-rwxr-xr-x 1 radio radio   21 May 10 13:32 t.liq
radio@sxvx:~$ 
```

Very interesting indeed. Can't explain why it works in 22.04 from scratch and dockerized LS can use /home/radio directory but not in 24.04.

**Reply by gAlleb** [reply]
There are some disadvantages of being humanitarian:) Now I'm learning about docker, args, Dockerfile and stuff.  Acl seems "distant")) But why, why it worked on 22.04 seemlessly. It just had access to /home/radio from the start?

**Reply by vitoyucepi** [reply]
Nope, nothing changed.
```
  File: /home/radio
Access: (0750/drwxr-x---)  Uid: ( 1000/ UNKNOWN)   Gid: ( 1000/ UNKNOWN)
```
In both ubuntu:22.04 and ubuntu:24.04 the default permissions for the home dir are `0750`, and the first `uid:gid` is `1000:1000`. In the `savonet/liquidsoap` container you are `uid=10000(liquidsoap) gid=10001(liquidsoap) groups=10001(liquidsoap),29(audio)`.

<details>
<summary>offtop</summary>

> Now I'm learning about docker, args, Dockerfile and stuff.

Generally, not everyone in the company will or should know about containers, file systems, Linux, you name it, even in the company staffed with programmers.

</details>

**Reply by gAlleb** [reply]
Thank you very much, @vitoyucepi! I'm learning. Also my level of knowledge was enough for 22.04 cause it worked there:) No Idea why it refuses to work on ubuntu 24.04. I've made the same setup with _second_ `radio` user (/etc/passwd shows `radio:x:1001:1001:,,,:/home/radio:/bin/bash`) same as in 22.04.

Probably I'm missing here something simple but important.

I've been thinking about converting opam setup into docker
1. Ubuntu 24.04. Docker 26.1.2.
2. *.m3u files are in `/home/radio/playlist` dir.
3. Music is mounted at `/mnt/GoogleDrive`
4. Local music `(large files and jingles)` is in `/home/radio/music`
5. HLS outputs to `/var/www/html/omfm/hls`
6. 8007 port

```ruby
services:
  liquidsoap:
    image: savonet/liquidsoap:v2.2.5
    command:
      - /home/radio/liquidsoap/1/index.liq
    ports:
      - 8007:8007
    volumes:
      - ./:/home/radio
      - /mnt/GoogleDrive:/mnt/GoogleDrive
      - /var/www/html/omfm/hls:/var/www/html/omfm/hls

```

Gonna study more but at the moment `the situation` really confuses me, cause `docker` worked fine in 22.04. Gonna dive further.

Thank  you.

**Reply by vitoyucepi** [reply]
Use `stat` to inspect the ownership and permissions.

**Reply by gAlleb** [reply]
```
On 24.04 
radio@sxvx:~$ stat /home/radio
  File: /home/radio
  Size: 4096      	Blocks: 8          IO Block: 4096   directory
Device: 253,2	Inode: 393255      Links: 14
Access: (0750/drwxr-x---)  Uid: ( 1001/   radio)   Gid: ( 1001/   radio)
Access: 2024-05-10 03:01:19.898441781 +0300
Modify: 2024-05-07 21:55:25.935142424 +0300
Change: 2024-05-07 21:55:25.935142424 +0300
 Birth: 2024-05-06 19:39:33.518124928 +0300

on 22.04


radio@sxvx:~$ stat /home/radio
  File: /home/radio
  Size: 4096      	Blocks: 8          IO Block: 4096   directory
Device: fc01h/64513d	Inode: 538         Links: 15
Access: (0755/drwxr-xr-x)  Uid: ( 1001/   radio)   Gid: ( 1002/   radio)
Access: 2024-05-06 03:01:16.592862438 +0300
Modify: 2024-04-29 21:27:45.611359277 +0300
Change: 2024-04-29 21:27:45.611359277 +0300
 Birth: 2023-03-05 12:28:55.705156413 +0300
```
 
GID are different and access, hm

**Reply by vitoyucepi** [reply]
The one with `0750` is obviously from *24.04* and the other one with `0755` is from *22.04*.

**Reply by gAlleb** [reply]
my bad. true

24.04
```ruby
radio@sxvx:~$ stat /home/radio
  File: /home/radio
  Size: 4096      	Blocks: 8          IO Block: 4096   directory
Device: 253,2	Inode: 393255      Links: 14
Access: (0750/drwxr-x---)  Uid: ( 1001/   radio)   Gid: ( 1001/   radio)
Access: 2024-05-10 03:01:19.898441781 +0300
Modify: 2024-05-07 21:55:25.935142424 +0300
Change: 2024-05-07 21:55:25.935142424 +0300
 Birth: 2024-05-06 19:39:33.518124928 +0300
```

**Reply by gAlleb** [reply]
So obviously I've missed something?

All I did is
```
adduser radio
usermod -aG sudo radio
rsync --archive --chown=radio:radio ~/.ssh /home/radio
```

But now it differs from 22.04. In 22.04 - it's `0755/drwxr-xr-x`. 

I presume that I'm dumb, so what's up?

**Reply by vitoyucepi** [reply]
The default permissions for the user home directory are `0750`, sometimes `0700`.

**Reply by gAlleb** [reply]
So its on `22.04` I've messed something up and it became `0755` and I could use docker within it?

**Reply by gAlleb** [reply]
Ok. Now I get it. I've did `~$ sudo chmod 0755 -R /home/radio` and LS docker started. So it's a mistake from dumb ubuntu 22.04 time:) Probably I've changed permissions.

But I presume - it's not right to give 755 to /home/user. Is it ok to give 755 to `liquidsoap` folder and `docker compose up` from there?

Also, Thanks, Vito! For navigating me through my stupidity.
